maven-core/src/main/java/org/apache/maven/plugin/internal/AbstractMavenPluginParametersValidator.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.maven.shared.utils.logging.MessageUtils;
+import org.codehaus.plexus.component.configurator.expression.ExpressionEvaluationException;
+import org.codehaus.plexus.component.configurator.expression.ExpressionEvaluator;
+import org.codehaus.plexus.configuration.PlexusConfiguration;
+import org.slf4j.Logger;
+
+/**
+ * Common implementations for plugin parameters configuration validation.
+ *
+ * @author Slawomir Jaranowski
+ */
+abstract class AbstractMavenPluginParametersValidator implements MavenPluginConfigurationValidator
+{
+
+    // plugin author can provide @Parameter( property = "session" ) in this case property will always evaluate
+    // so, we need ignore those
+
+    // source org.apache.maven.plugin.PluginParameterExpressionEvaluator
+    private static final List<String> IGNORED_PROPERTY_VALUES = Arrays.asList(
+        "basedir",
+        "executedProject",
+        "localRepository",
+        "mojo",
+        "mojoExecution",
+        "plugin",
+        "project",
+        "reactorProjects",
+        "session",
+        "settings"
+    );
+
+    private static final List<String> IGNORED_PROPERTY_PREFIX = Arrays.asList(
+        "mojo.",
+        "plugin.",
+        "pom.",
+        "project.",
+        "session.",
+        "settings."
+    );
+
+    protected abstract Logger getLogger();
+
+    protected static boolean isValueSet( PlexusConfiguration config,
+                                         ExpressionEvaluator expressionEvaluator )
+    {
+        if ( config == null )
+        {
+            return false;
+        }
+
+        // there are sub items ... so configuration is declared
+        if ( config.getChildCount() > 0 )
+        {
+            return true;
+        }
+
+        String strValue = config.getValue();
+
+        if ( strValue == null || strValue.isEmpty() )
+        {
+            return false;
+        }
+
+        if ( isIgnoredProperty( strValue ) )
+        {
+            return false;
+        }
+
+        // for declaration like @Parameter( property = "config.property" )
+        // the value will contains ${config.property}
+
+        try
+        {
+            return expressionEvaluator.evaluate( strValue ) != null;
+        }
+        catch ( ExpressionEvaluationException e )
+        {
+            // not important
+            // will be reported during Mojo fields populate
+        }
+
+        // fallback - in case of error in expressionEvaluator
+        return false;
+    }
+
+    private static boolean isIgnoredProperty( String strValue )
+    {
+        if ( !strValue.startsWith( "${" ) )
+        {
+            return false;
+        }
+
+        String propertyName = strValue.replace( "${", "" ).replace( "}", "" );
+
+        if ( IGNORED_PROPERTY_VALUES.contains( propertyName ) )
+        {
+            return true;
+        }
+
+        return IGNORED_PROPERTY_PREFIX.stream().anyMatch( propertyName::startsWith );
```

Comment:
Wow, this is sick :-D

Reply:
lambda way ... what is your proposition instead of ...

Reply:
No, nope. I just find the expression extremely cool.